diff --git a/lib/kernel/src/inet_db.erl b/lib/kernel/src/inet_db.erl
index 814c3f4276..b33a07294b 100644
--- a/lib/kernel/src/inet_db.erl
+++ b/lib/kernel/src/inet_db.erl
@@ -635,6 +635,14 @@ lookup_type(Domain, Type) ->
 lookup_cname(Domain) ->
     [R#dns_rr.data || R <- lookup_rr(Domain, in, ?S_CNAME) ].
 
+lookup_cname(Domain, Type) ->
+    case Type of
+     a -> [];
+     aaaa -> [];
+     cname -> lookup_cname(Domain);
+     _ -> []
+    end.
+
 %% Have to do all lookups (changes to the db) in the
 %% process in order to make it possible to refresh the cache.
 lookup_rr(Domain, Class, Type) ->
@@ -710,9 +718,10 @@ ent_gethostbyaddr(RRs, IP, AddrType, Length) ->
 		    ?dbg("gethostbyaddr found extra=~p~n", [TR]);
 	       true -> ok
 	    end,
+            Type = RR#dns_rr.type,
 	    Domain = RR#dns_rr.data,
 	    H = #hostent { h_name = Domain,
-			   h_aliases = lookup_cname(Domain),
+			   h_aliases = lookup_cname(Domain, Type),
 			   h_addr_list = [IP],
 			   h_addrtype = AddrType,
 			   h_length = Length },
